ipmininet/ipnet.py

```diff
@@ -549,6 +550,79 @@ def ping6Pair(self):
            return: ploss packet loss percentage"""
         return self.pingPair(use_v4=False)
 
+   
+    def runFailurePlan(self, failure_plan: List[Tuple[str,str]]) -> List[IPIntf]:
+        """this function run a failure plan
+            :param: A list of tuple of string giving the name of the 2 nodes between
+                    which the link have to be taken down
+            :return: A list of IPIntf which represent the interface that goes down or 
+                    an empty list if there was an unknown node in the failure_plan
+        """
+        log.output("** Starting failure plan\n")
+        interfaces_down = []
+        for link in failure_plan:
+            try:
+                node1 = self.get(link[0])
+                node2 = self.get(link[1])
+                interfaces = node1.connectionsTo(node2)
+                interfaceNode1 = interfaces[0][0]
+                interfaceNode2 = interfaces[0][1]
+                interfaces_down.append(interfaceNode1)
+            except KeyError as e:
+                log.output("ERROR: node "+ str(e) +"\n")
+                interfaces_down = []
+        for interface in interfaces_down:
+            commande = "ip link set dev " + str(interface) + " down"
+            interface.node.cmd(commande)
+            log.output("** Interface "+ str(interface) +" down\n")
+        return interfaces_down
+
+    def restoreLink(self, interfaces: List[IPIntf]):
+        """function which restore the link
+            :param interfaces: a List of IPIntf
+        """
+        log.output("** starting restoring link\n")
+        for interface in interfaces:
+            commande = "ip link set dev " + str(interface) + " up"
+            interface.node.cmd(commande)
+            log.output("** interfaces " + str(interface) + " up\n")
+
+    def RandomFailure(self, n: int, weak_links:List[IPIntf]=None) -> List[IPIntf]:
```

Comment:
You need to specify ``Optional`` if a parameter can be ``None``

```suggestion
    def RandomFailure(self, n: int, weak_links:Optional[List[IPIntf]]=None) -> List[IPIntf]:
```